src/main/java/sx/blah/discord/util/MessageHistoryBuilder.java

```diff
@@ -0,0 +1,429 @@
+/*
+ *     This file is part of Discord4J.
+ *
+ *     Discord4J is free software: you can redistribute it and/or modify
+ *     it under the terms of the GNU Lesser General Public License as published by
+ *     the Free Software Foundation, either version 3 of the License, or
+ *     (at your option) any later version.
+ *
+ *     Discord4J is distributed in the hope that it will be useful,
+ *     but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *     GNU Lesser General Public License for more details.
+ *
+ *     You should have received a copy of the GNU Lesser General Public License
+ *     along with Discord4J.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package sx.blah.discord.util;
+
+import sx.blah.discord.handle.impl.obj.Channel;
+import sx.blah.discord.handle.obj.IChannel;
+import sx.blah.discord.handle.obj.IMessage;
+
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static sx.blah.discord.handle.impl.obj.Channel.MESSAGE_CHUNK_COUNT;
+
+/**
+ * This is a utility class for building message histories in a readable, easy-to-understand fashion. It can be built
+ * into a {@link MessageHistory} object, a {@link Stream} of {@link IMessage} objects, or an {@link Iterator} of
+ * {@link IMessage} objects. It can also be directly used in a for-each loop without being built.
+ *
+ * @see MessageHistory
+ * @see MessageHistoryIterator
+ */
+public class MessageHistoryBuilder implements Iterable<IMessage> {
+
+	private Channel channel;
+
+	// User provided info
+	private IMessage uStart; // Null if not given
+	private LocalDateTime uStartTime; // null signifies most recent
+	private boolean uIncludeStart;
```

Comment:
Well-defined and documented behavior is superior to the mess caused by allowed either end to inclusive.